rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalTextDocumentService.java

```diff
@@ -542,6 +542,8 @@ public CompletableFuture<List<? extends CodeLens>> codeLens(CodeLensParams param
                 return r;
             })
             .thenApply(Versioned::get)
+            // Replace with the last tree without errors, might be the same as `tree` if the parse succeeded without any error recovery
+            .thenApply(tree -> f.getLastTreeWithoutErrors().get())
```

Comment:
This feels very temporal coupling to the semantics of `getLastTreeWithoutErrors` (also evidenced by the fact that some logic had to be switched over there).

Are we sure this is enough of a fix? Since looking at the `TextDocumentState` it looks more like the `getLastTreeWithoutErrors` is about parse exceptions and not about trees with error recovery or not.

Reply:
`lastTreeWithoutErrors` is only updated if no parse error or exception occured and the parse tree contains no errors. 

I think updating `lastWithoutErrors` before completing the `treeAsync` future is actually a sensible fix. Otherwise the `lastTreeWithoutErrors` premise is not true in the code executed on completion of the future.

I have already tested this fix and it seems to work great!